docs: describe workaround for missing CSRF token for reverse proxied Grafana (#6149)
sourcegraph/sourcegraph: doc/admin/monitoring_and_tracing.md

> NOTE: We are running Grafana behind a reverse proxy. Grafana is not fully integrated with our CSRF protection so there is a known issue: when the Grafana
> web app in the browser makes POST or PUT requests Sourcegraph's CSRF protection gets triggered and responds with a "invalid CSRF token" 403 response.
> We are working to solve [this issue](https://github.com/sourcegraph/sourcegraph/issues/6075). 
> As a workaround, site admins can connect to Grafana directly (as described below) to make changes to the dashboards. 

### Exposing Grafana directly

If you are running Sourcegraph as single server you can add an additional publish flag for the Grafana port 3370:

```shell script

> docker run --publish 7080:7080 --publish 2633:2633 --publish 3370:3370 \

Reviewer: In most server deployments, there isn't going to be any firewall configured so this is an inherently dangerous action. You would effectively be exposing Grafana with anonymous access directly to the public internet. We should call that out explicitly here and clarify that Grafana should only be exposed JUST in the case you're trying to workaround the bug mentioned above.

Reviewer: Alternatively, we could maybe provide a command to sshuttle proxy the Grafana port? i.e. sshuttle into the remote machine running Sourcegraph, then into the Docker container itself (I'm not sure how hard that would be to construct)

  --rm --volume ~/.sourcegraph/config:/etc/sourcegraph \
  --volume ~/.sourcegraph/data:/var/opt/sourcegraph sourcegraph/server:3.9.1
```

If you're using the [Kubernetes cluster deployment option](https://github.com/sourcegraph/deploy-sourcegraph),  
you can run Grafana directly using kubectl port-forwarding

```shell script
kubectl port-forward svc/grafana 3370:30070
``` 

Then visit `http://localhost:3370/-/debug/grafana` to get to the "Home Dashboard". From there you can add, modify and delete dashboards and panels.

### Docker images

#### Prometheus

We are running our own image of Prometheus which contains a standard Prometheus installation packaged together 
with rules files and target files for our monitoring.

A directory can be mounted at `/sg_prometheus_add_ons`. It can contains additional config files of two types:
  - rule files which must have the suffix `_rules.yml` in their filename (ie `gitserver_rules.yml`)
  - target files which must have the suffix `_targets.yml` in their filename (ie `local_targets.yml`)

[Rule files](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/) 

> and [target files](https://prometheus.io/docs/guides/file-sd/) need to use Prometheus 2.x syntax.  

Reviewer: ```suggestion
and [target files](https://prometheus.io/docs/guides/file-sd/) must use the latest Prometheus 2.x syntax.  
```

> NOTE: We are running Grafana behind a reverse proxy. Grafana is not fully integrated with our CSRF protection so there is a known issue: when the Grafana
> web app in the browser makes POST or PUT requests Sourcegraph's CSRF protection gets triggered and responds with a "invalid CSRF token" 403 response.
> We are working to solve [this issue](https://github.com/sourcegraph/sourcegraph/issues/6075). 
> As a workaround, site admins can connect to Grafana directly (as described below) to make changes to the dashboards. 

> ### Exposing Grafana directly

Reviewer: Instead of one big section, could you create three subsections below this one?

> #### Kubernetes
> ...
>
> #### Single-container server deployments
> ...
>
> #### Single-container server deployments (with firewall)
> ...

Reviewer: Also, change `Exposing Grafana directly` to `Accessing Grafana directly` since most options here aren't actually exposing it

Author: ok

> If you are running Sourcegraph as single server behind a firewall you can add an additional publish flag for the Grafana port 3370:

Reviewer: I would move this further down, below the non-firewall version. i.e. we assume the worst by default (no firewall) and then for "advanced users" offer the command to expose it to the public if they have a firewall

Author: ok

```shell script
docker run --publish 7080:7080 --publish 2633:2633 --publish 3370:3370 \
  --rm --volume ~/.sourcegraph/config:/etc/sourcegraph \
  --volume ~/.sourcegraph/data:/var/opt/sourcegraph sourcegraph/server:3.9.1
```

If you are running Sourcegraph as a single server in a droplet or otherwise exposed to the open internet then publishing
the additional port is not recommended since the embedded Grafana runs in anonymous mode. In this case we suggest using
the tools `sshuttle` and `socat` to create a secure connection. Log into your remote server using ssh 
and run these commands: 

```shell script
docker exec -t XXX_CONTAINER_ID_XXXX apk add --no-cache socat
socat tcp-listen:3370,reuseaddr,fork system:"docker exec -i 7298cae012eb socat stdio 'tcp:localhost:3370'"
```

On your local machine start a `sshuttle` session to your remote server

```shell script
sshuttle -r xxx_user_xxx@sxxx_remote_server_xxx 0/0
```

If you're using the [Kubernetes cluster deployment option](https://github.com/sourcegraph/deploy-sourcegraph),  

> you can run Grafana directly using kubectl port-forwarding

Reviewer: ```suggestion
you can access Grafana directly using Kubernetes port forwarding to your local machine:
```

```shell script
kubectl port-forward svc/grafana 3370:30070
``` 

> Then visit `http://localhost:3370/-/debug/grafana` to get to the "Home Dashboard". From there you can add, modify and delete dashboards and panels.

Reviewer: ```suggestion
Then visiting http://localhost:3370/-/debug/grafana will show the home dashboard and from there you can add, modify and delete your own dashboards and panels, and configure alerts.
```
